### PR TITLE
#35048 Add support for public schemas for semantic query analysis

### LINKED ISSUE #34807: New autocomplete mode doesn't suggest system tables
### Description

New autocomplete mode doesn't suggest system tables, for example, pg_class.

Start typing: select * from pg_c
press Ctrl + Enter

Expected: "pg_class" is in a list of suggestions
Actual: No suggestions at all

### DBeaver Version

Community Edition 24.1.3.202407122322

### Operating System

Windows 10

### Database and driver

PostgreSQL 16
PostgreSQL JDBC Driver 42.7.2

### Steps to reproduce

_No response_

### Additional context

_No response_

### LINKED ISSUE #35048: System table names without specifying optional schema name are colored with wrong color
### Description

Both queries are executed correctly but "pg_class" in the second query is colored with red color what means semantic error.
Use this code:
```
select * from pg_catalog.pg_class

select * from pg_class as pc
```

изображение

It is similar to this: #34741

### DBeaver Version

Community Edition 24.1.4.202407250946

### Operating System

Windows 10

### Database and driver

PostgreSQL 16
PostgreSQL JDBC Driver 42.7.2

### LINKED ISSUE #35430: INFORMATION_SCHEMA views aren't recognized in SQL Server
### Description

When trying to use INFORMATION_SCHEMA in SQL Server, the views are marked as not found.  Such as:

`SELECT *
FROM INFORMATION_SCHEMA.ROUTINES`

I get a message _Table INFORMATION_SCHEMA.ROUTINES not found_, like:

SELECT *
_Table INFORMATION_SCHEMA.ROUTINES not found_
FROM INFORMATION_SCHEMA._ROUTINES_

If I just have

SELECT *
FROM INFORMATION_SCHEMA

and don't type fast enough, I get 
_Table INFORMATION_SCHEMA not found_

The new analysis is nice, and it's getting better!, but still has some issues.  <wry smile>

[dbeaver-diagnostic-info-1725285105290.zip](https://github.com/user-attachments/files/16838713/dbeaver-diagnostic-info-1725285105290.zip)

### DBeaver Version

Community edition 24.2.0

### Operating System

Windows 10

### Database and driver

Microsoft JDBC Driver for SQL Server 
12.8.0.jre11

### Steps to reproduce

1. Connect to SQL Server
2. Type in 

SELECT *
FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE

and wait for the error.  <wry smile>

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by E1izabeth**
### For QA team: please pay attention, there are 3 linked issues

PostgreSQL
```sql
-- simple table in custom schema, which is current execution context

select * from "Employee"  -- works, hilighted correctly

select * from "Elizabeth"."Employee" -- works, highlighted correctly

-- simple table in "public" schema of the current current execution context's database

select * from "TicketQueue"  -- works, BUT TABLE NOT FOUND HIGHLIGHTED

select * from public."TicketQueue" -- works, highlighted correctly

select * from "Elizabeth"."TicketQueue" -- doesn't work highlighted correctly

-- information_schema view 

select * from columns;  -- doesn't work, BUT HIGHLIGHTED AS OK

select * from information_schema.columns -- works, hilighted correctly

select * from "Elizabeth".columns -- doesn't work highlighted correctly

-- system catalog

select * from pg_class  -- works, BUT TABLE NOT FOUND HIGHLIGHTED

select * from pg_catalog.pg_class -- works, highlighted correctly

select * from "Elizabeth".pg_class -- doesn't work highlighted correctly

```

Oracle
```sql
SELECT * FROM elizabeth.yuuy  ; -- synonym for employee table 

SELECT * FROM "PUBLIC".customers ; -- public synonym for table employee


-- local table in the current execution context

select * from SALES; -- works

select * from Elizabeth.sales; -- works

-- SYS

select * from "TAB$" -- doesn't work

select * from SYS."TAB$" -- works

select * from Elizabeth."TAB$" -- doesn't work

-- SYSTEM

select * from REDO_DB -- doesn't work

select * from system.REDO_DB -- works

select * from Elizabeth.REDO_DB -- doesn't work

select * from SYS.REDO_DB -- doesn't work

```

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/struct/DBSVisibilityScopeProvider.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * Provides additional globally visible scopes
+ */
+public interface DBSVisibilityScopeProvider {
+    /**
+     * Returns additional globally visible scopes, like information_schema
+     */
+    default List<DBSObjectContainer> getPublicScopes(@NotNull DBRProgressMonitor monitor) throws DBException {
+        return Collections.emptyList();
+    }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/struct/DBSVisibilityScopeProvider.java]: Does it really need a default implementation? It seems all its current implementors have it implemented.

File: plugins/org.jkiss.dbeaver.ext.sqlite/src/org/jkiss/dbeaver/ext/sqlite/model/SQLiteTable.java [hunk 2]
```diff
@@ -81,54 +82,54 @@ protected boolean isTruncateSupported() {
...
-                    .map(a -> a.getName().toLowerCase())
-                    .collect(Collectors.toSet());
-                this.allPseudoAttributes = ALL_KNOWN_PSEUDO_ATTRS.stream()
-                    .filter(a -> !columnNames.contains(a.getName())) // all names are lowercased here
-                    .toArray(DBDPseudoAttribute[]::new);
-            }
+        return this.allPseudoAttributes != null ? this.allPseudoAttributes : (this.allPseudoAttributes = obtainAllPseudoAttributes(this.getDataSource(), this, monitor));
+    }
+
+    public static DBDPseudoAttribute[] obtainAllPseudoAttributes(GenericDataSource dataSource, DBSTable table, @NotNull DBRProgressMonitor monitor) throws DBException {
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ext.sqlite/src/org/jkiss/dbeaver/ext/sqlite/model/SQLiteTable.java]: Monitor must be the first parameter

File: plugins/org.jkiss.dbeaver.ext.sqlite/src/org/jkiss/dbeaver/ext/sqlite/model/SQLiteTable.java [hunk 3]
```diff
@@ -81,54 +82,54 @@ protected boolean isTruncateSupported() {
...
+        boolean isWithoutRowId = obtainIsWithoutRowId(dataSource, table, monitor);
+        if (isWithoutRowId) {
+            return DBDPseudoAttribute.EMPTY_ARRAY;
+        } else {
+            // see https://www.sqlite.org/lang_createtable.html#rowid (5. ROWIDs and the INTEGER PRIMARY KEY):
+            //     If a table contains a user defined column named "rowid", "oid" or "_rowid_",
+            //     then that name always refers the explicitly declared column and cannot be used to retrieve the integer rowid value.
+
+            Set<String> columnNames = table.getAttributes(monitor).stream()
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ext.sqlite/src/org/jkiss/dbeaver/ext/sqlite/model/SQLiteTable.java]: Two new sets + case conversion, seems to be over-complicated.
Can you just return ALL_KNOWN_PSEUDO_ATTRS filtered with column presence check? Like table.getAttribute(monitor, name)?

File: plugins/org.jkiss.dbeaver.ext.sqlite/src/org/jkiss/dbeaver/ext/sqlite/model/SQLiteTable.java [hunk 4]
```diff
@@ -156,8 +157,8 @@ private boolean obtainIsWithoutRowId(@NotNull DBRProgressMonitor monitor) throws
         }
     }
 
-    private boolean hasPrimaryKey() throws DBException {
-        List<GenericUniqueKey> constraints = getConstraints(new VoidProgressMonitor());
+    public static boolean hasPrimaryKey(DBSTable table) throws DBException {
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ext.sqlite/src/org/jkiss/dbeaver/ext/sqlite/model/SQLiteTable.java]: Also an over-complicated check. Do we really need it?
Can we just return pseudo attrs like they are always there?

**Review by serge-rider — CHANGES_REQUESTED** [review]

**Comment by uslss** [comment]
verified

**Pull request merged**